libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -80,20 +80,35 @@ string YulUtilFunctions::splitExternalFunctionIdFunction()
 	});
 }
 
-string YulUtilFunctions::copyToMemoryFunction(bool _fromCalldata)
+string YulUtilFunctions::copyToMemoryFunction(bool _fromCalldata, bool _cleanup)
 {
-	string functionName = "copy_" + string(_fromCalldata ? "calldata" : "memory") + "_to_memory";
+	string functionName =
+		"copy_" +
+		string(_fromCalldata ? "calldata" : "memory") +
+		"_to_memory" +
+		string(_cleanup ? "_cleanup" : "");
+
 	return m_functionCollector.createFunction(functionName, [&]() {
 		if (_fromCalldata)
 		{
 			return Whiskers(R"(
 				function <functionName>(src, dst, length) {
 					calldatacopy(dst, src, length)
-					// clear end
-					mstore(add(dst, length), 0)
+					<?cleanup>
+					// clear up to 32 byte boundary
+					let endPos := add(dst, length)
+					let spill := and(endPos, 0x1F)
+					if spill
+					{
+						let mask := <shl>(not(0), sub(0x20, spill))
+						mstore(endPos, and(mload(endPos), mask))
+					}
```

Comment:
Out of curiosity: Have you checked if something fanciful and branchless like
```suggestion
					// clear up to 32 byte boundary
					calldatacopy(add(dst, length), calldatasize(), and(sub(32, length), 31))
```
is cheaper?

Reply:
[Resp. from our discussion in chat we might want to refert back to the original ``mstore(add(dst, length), 0)`` anyways]